### PR TITLE
Abstract some ec2 CI logic and add support for c7g

### PR DESCRIPTION
**Pull request by samuel40791765**
### Issues:
Resolves `CryptoAlg-1121`

### Description of changes: 
We're looking to strengthen our CI coverage. This extends the M1 Mac CI framework to encapsulate specific ec2 instances we want to test against. Things are slightly different than M1 however. M1 instances need to be allocated for at least 24 hours, while these instances can be spun up and terminated on the fly.

### Call-outs:
This, again, borrows a lot of code from our original benchmarking framework. Scripts and SSM document were taken from:
* https://github.com/aws/aws-lc/blob/main/tests/ci/run_bm_framework.sh
* https://github.com/aws/aws-lc/blob/main/tests/ci/cdk/cdk/ssm/bm_framework_ssm_document.yaml

### Testing:
CI changes are tested and can be seen in a PR on my local fork:
* https://github.com/samuel40791765/aws-lc/pull/26
See:  AWS CodeBuild BuildBatch us-west-2 (aws-lc-ci-ec2-test-framework) 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

**Comment by samuel40791765** [comment]
I changed the name of "aws-lc-ci-macos-arm" to "aws-lc-ci-ec2-test-framework". That build batch won't be kicked off until we apply this change :).

File: tests/ci/cdk/cdk/aws_lc_ec2_test_framework_ci_stack.py [hunk 1]
```diff
@@ -97,18 +96,18 @@ def __init__(self,
         security_group = ec2.SecurityGroup(self, id="{}-ec2-sg".format(id),
                           allow_all_outbound=True,
                           vpc=vpc,
-                          security_group_name='macos_arm_ec2_sg')
+                          security_group_name='codebuild_ec2_sg')
```
Comment (billbo-yang) [inline tests/ci/cdk/cdk/aws_lc_ec2_test_framework_ci_stack.py:99]: What's the difference between `macos_arm_ec2_sg` and `codebuild_ec2_sg`?

Comment (samuel40791765) [reply tests/ci/cdk/cdk/aws_lc_ec2_test_framework_ci_stack.py:99]: No difference, it was part of making the ec2 logic more applicable to all instances. `codebuild_ec2_sg` is now used for the graviton3 instance as well, so the original name didn't make much sense anymore.

**Review by billbo-yang — APPROVED** [review]

File: tests/ci/cdk/cdk/ssm/general_test_run_ssm_document.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        # install docker if its not already installed
+        - chmod +x ./tests/ci/benchmark_framework/install_docker.sh
+        - ./tests/ci/benchmark_framework/install_docker.sh
+        # log into docker and get needed docker image from ecr
+        - export ECR_REPO="{AWS_ACCOUNT_ID}.dkr.ecr.us-west-2.amazonaws.com/aws-lc-docker-images-${CPU_ARCH}"
+        - aws ecr get-login-password --region us-west-2 | docker login -u AWS --password-stdin "${ECR_REPO}"
+        - docker pull "${ECR_REPO}:amazonlinux-2023_clang-15x_sanitizer"
+        - # start the container and run the bm script
+        - exec_docker="docker run --env AWS_ACCOUNT_ID={AWS_ACCOUNT_ID} --env PR_NUM={PR_NUM} --env COMMIT_ID={COMMIT_ID} --env CPU_TYPE=${CPU_TYPE} -v `pwd`:`pwd` -w `pwd` ${ECR_REPO}:amazonlinux-2023_clang-15x_sanitizer_latest"
+        - chmod +x ./tests/ci/run_posix_sanitizers.sh
```
Comment (skmcgrail) [inline tests/ci/cdk/cdk/ssm/general_test_run_ssm_document.yaml:37]: Is this not already marked as executable in the repository?

File: tests/ci/cdk/cdk/ssm/general_test_run_ssm_document.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        - git clone {GITHUB_REPO} aws-lc-pr
+        - cd aws-lc-pr
+        - git checkout {COMMIT_ID}
+        # install docker if its not already installed
+        - chmod +x ./tests/ci/benchmark_framework/install_docker.sh
+        - ./tests/ci/benchmark_framework/install_docker.sh
+        # log into docker and get needed docker image from ecr
+        - export ECR_REPO="{AWS_ACCOUNT_ID}.dkr.ecr.us-west-2.amazonaws.com/aws-lc-docker-images-${CPU_ARCH}"
+        - aws ecr get-login-password --region us-west-2 | docker login -u AWS --password-stdin "${ECR_REPO}"
+        - docker pull "${ECR_REPO}:amazonlinux-2023_clang-15x_sanitizer"
```
Comment (skmcgrail) [inline tests/ci/cdk/cdk/ssm/general_test_run_ssm_document.yaml:34]: Probably make the image name a variable, as it is repeated below in the exec_docker definition.

File: tests/ci/cdk/app.py [hunk 4]
```diff
@@ -50,8 +50,8 @@
 AwsLcGitHubAnalyticsStack(app, "aws-lc-ci-analytics", analytics_build_spec_file, env=env)
 # bm_framework_build_spec_file = "cdk/codebuild/bm_framework_omnibus.yaml"
 # BmFrameworkStack(app, "aws-lc-ci-bm-framework", bm_framework_build_spec_file, env=env)
-mac_arm_build_spec_file = "cdk/codebuild/github_ci_macos_m1_omnibus.yaml"
-AwsLcMacArmCIStack(app, "aws-lc-ci-macos-arm", mac_arm_build_spec_file, env=env)
+ec2_test_framework_build_spec_file = "cdk/codebuild/ec2_test_framework_omnibus.yaml"
```
Comment (skmcgrail) [inline tests/ci/cdk/app.py:53]: I can't remember, but I think we will need to remember to manually terminate this stack. It looks like the resources created in it are all unique I believe? Don't believe CDK will handle this for us in this instance.

File: tests/ci/run_ec2_test_framework.sh [hunk 5]
```diff
@@ -0,0 +1,132 @@
+#!/bin/bash
```
Comment (skmcgrail) [inline tests/ci/run_ec2_test_framework.sh:1]: suggestion: Probably not consistent elsewhere in our repo, but I tend to like the shebang to be something like `#!/usr/bin/env bash`. Lets the environment control which specific bash command to use by letting `env` search `PATH` for you. It's available in mac as well, and have yet to find a system image that doesn't provide it in a default system. (Likely cause it is part of the GNU coreutils).

Comment (samuel40791765) [reply tests/ci/run_ec2_test_framework.sh]: Sure I just took this from other files in our repo, but it's definately great to have a standard we want to enforce.

**Review by skmcgrail — APPROVED** [review]

**Review by billbo-yang — APPROVED** [review]

**Pull request merged**